gremlin/ui/device.py

```diff
@@ -30,13 +31,108 @@
 from gremlin import common, event_handler, joystick_handling, shared_state
 from gremlin.error import GremlinError
 from gremlin.intermediate_output import IntermediateOutput
-from gremlin.types import InputType
+from gremlin.types import InputType, PropertyType
+from gremlin.util import file_exists_and_is_accessible
+from gremlin.common import SingletonDecorator
+from gremlin.config import Configuration
 
 
 QML_IMPORT_NAME = "Gremlin.Device"
 QML_IMPORT_MAJOR_VERSION = 1
 
 
+@SingletonDecorator
+class DeviceDatabase:
+    """Provides button/axis to name mapping for known devices"""
+
+    def __init__(self) -> None:
+        self._load()
+
+    def _load(self) -> None:
+
+        db_file = "device_db.json"
+        if not file_exists_and_is_accessible(db_file):
+            return
+
+        load_successful = False
+        json_data = {}
+        with open(db_file) as hdl:
+            try:
+                decoder = json.JSONDecoder()
+                json_data = decoder.decode(hdl.read())
+                load_successful = True
+            except ValueError:
+                pass
+
+        if not load_successful:
+            logging.getLogger("system").error(
+                    f"There was an error loading {db_file}. "
+                    "Probably it's a syntax error, "
+                    "please fix the errors and restart the application.")
+
+        # create an empty device database if load was not successful
+        self.device_db = json_data if load_successful else {}
```

Comment:
Does this not having "device" and "mapping" keys cause issues?

Reply:
Good catch. Added defaults.